### PR TITLE
[SPARK-6752][Streaming] Allow StreamingContext to be recreated from checkpoint and existing SparkContext

### PR DESCRIPTION
**Pull request by tdas**
Currently if you want to create a StreamingContext from checkpoint information, the system will create a new SparkContext. This prevent StreamingContext to be recreated from checkpoints in managed environments where SparkContext is precreated.

The solution in this PR: Introduce the following methods on StreamingContext
1. `new StreamingContext(checkpointDirectory, sparkContext)`
   Recreate StreamingContext from checkpoint using the provided SparkContext
2. `StreamingContext.getOrCreate(checkpointDirectory, sparkContext, createFunction: SparkContext => StreamingContext)`
   If checkpoint file exists, then recreate StreamingContext using the provided SparkContext (that is, call 1.), else create StreamingContext using the provided createFunction

TODO: the corresponding Java and Python API has to be added as well.

### CODE REVIEW

**Comment by tdas** [comment]
@jerryshao Mind taking a look at this? Its still WIP as unit tests are commented out.

**Comment by zzcclp** [comment]
@tdas , can this RP resolve [this issue](https://issues.apache.org/jira/browse/SPARK-5206)? 
Restart a streaming app from checkpoint  incorrectly if using accumulators .

**Comment by jerryshao** [comment]
@tdas Yeah, will do.

@zzcclp I'm not sure, maybe you can take a try, from my guess, this could possibly work, since accumulator is registered in SparkContext, while SparkContext is still existed.

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala [hunk 1]
```diff
@@ -271,7 +282,10 @@ object CheckpointReader extends Logging {
     })
 
     // If none of checkpoint files could be read, then throw exception
-    throw new SparkException("Failed to read checkpoint from directory " + checkpointPath)
+    if (!ignoreReadError) {
+      throw new SparkException("Failed to read checkpoint from directory " + checkpointPath)
```
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: Is it better to change to string interpolator style?

**Comment by jerryshao** [comment]
It looks good to me. Simply curious about the scenarios of this usage, is there any situation where streaming context is failed but spark context is still existed when driver failure is met?

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala [hunk 2]
```diff
@@ -77,7 +77,8 @@ object Checkpoint extends Logging {
   }
 
   /** Get checkpoint files present in the give directory, ordered by oldest-first */
-  def getCheckpointFiles(checkpointDir: String, fs: FileSystem): Seq[Path] = {
+  def getCheckpointFiles(checkpointDir: String, fsOption: Option[FileSystem] = None): Seq[Path] = {
```
Comment (harishreedharan) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: This change seems unrelated to this fix. I think we can simply do a null check inside this method and create a FileSystem if needed to avoid unnecessary changes to the calls (all the fs being passed in changing to Some(fs)) -- keeps git history sane.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 3]
```diff
@@ -114,11 +123,15 @@ class StreamingContext private[streaming] (
 
   private[streaming] val isCheckpointPresent = (cp_ != null)
 
+  private[streaming] val isSparkContextPresent = (sc_ != null)
```
Comment (harishreedharan) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: This seems to have been defined for setting `sc` below, but once `sc` is set this variable can be false, while `sc` is no longer `null`. This `val` should probably not be added, else we risk ending up having bugs caused by this being `false` which it can never be, since a spark context is always present.

**Comment by zzcclp** [comment]
@jerryshao , thanks, I will test this case later.

**Comment by tdas** [comment]
@zzcclp I dont think it will solve this issue directly. But it may allow the SparkContext to be re-initialized properly before the StreamingContext is recreated from checkpoints.

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala (hunk 1 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: Yes, I will. Thanks!

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala (hunk 2 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: The reason I added this is so that we should not have to handle nulls. Dealing with nulls is severely frowned upon in Sclaa, and precisely why Option was introduced. There are many places where this has been done, and slowly I was fix those. I think this is a small enough change (doesnt change functionality, or existing code paths) that is okay to do this.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala (hunk 3 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Good point. My idea was to keep it consistent with isCheckpointPresent, which has the semantics of "was checkpoint present at the time of creation of the StremaingContext". So this was meant to be "was an existing sparkcontext used to create the stremaingContext". But i can see that it can be confusing. Will remove that completely.

**Comment by tdas** [comment]
@all This is still a WIP. Adding the equivalent Java API requires refactoring the existing `JavaStreamingContext.getOrCreate` to not use `JavaStreamingContextFactory` and use `o.a.s.java.api.Function0` (which needs to be added). I am going to open other PRs for them before this can be merged.

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala (hunk 2 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: BTW, this file has to change anyways in the attempt to make the semantics of `read` more clear.

**Comment by tdas** [comment]
@JoshRosen Please take a quick look at the Function. 
@jerryshao @harishreedharan I have updated the patch with Java API and unit tests. I think I am going to create a separate JIRA for python API.

File: streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java [hunk 4]
```diff
@@ -987,12 +988,12 @@ public void testPairMap2() { // Maps pair -> single
     JavaDStream<Tuple2<String, Integer>> stream = JavaTestUtils.attachTestInputStream(ssc, inputData, 1);
     JavaPairDStream<String, Integer> pairStream = JavaPairDStream.fromJavaDStream(stream);
     JavaDStream<Integer> reversed = pairStream.map(
```
Comment (tdas) [inline streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java]: Some of these are unrelated to the PR, but just cleans up the formatting of the JavaAPISuite which is quite badly formatted.

**Comment by tdas** [comment]
Jenkins, test this again.

**Comment by tdas** [comment]
Jenkins, test this.

**Comment by zzcclp** [comment]
@tdas , I tested streaming recovering from checkpoint with this PR, it failed if it use accumulators, so this assuredly can't solve [issue SPARK-5206](https://issues.apache.org/jira/browse/SPARK-5206) directly. how to solve [issue SPARK-5206](https://issues.apache.org/jira/browse/SPARK-5206)?

**Comment by tdas** [comment]
Yes, this is not intended to solve SPARK-5206.

File: streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala (hunk 1 above)
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/Checkpoint.scala]: Do you think that we should log a warning message in the case where we ignore the error?

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 5]
```diff
@@ -107,6 +107,15 @@ class StreamingContext private[streaming] (
    */
   def this(path: String) = this(path, new Configuration)
 
+
+  def this(path: String, sparkContext: SparkContext) = {
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: This should probably have scaladoc to explain that it restores from a checkpoint?

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala [hunk 6]
```diff
@@ -621,19 +636,59 @@ object StreamingContext extends Logging {
...
+   * recreated from the checkpoint data. If the data does not exist, then the StreamingContext
+   * will be created by called the provided `creatingFunc` on the provided `sparkContext`. Note
+   * that the SparkConf configuration in the checkpoint data will not be restored as the
+   * SparkContext has already been created.
+   *
+   * @param checkpointPath Checkpoint directory used in an earlier StreamingContext program
+   * @param creatingFunc   Function to create a new StreamingContext using the given SparkContext
+   * @param sparkContext   SparkContext using which the StreamingContext will be created
+   */
+  def getOrCreate(
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Do you need this particular overloaded method?  Couldn't you just provide `false` as a default argument to `createOnError` in the next line?

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala [hunk 7]
```diff
@@ -655,6 +656,7 @@ object JavaStreamingContext {
    * @param checkpointPath Checkpoint directory used in an earlier JavaStreamingContext program
    * @param factory        JavaStreamingContextFactory object to create a new JavaStreamingContext
    */
+  @deprecated("use getOrCreate without JavaStreamingContextFactor", "1.4.0")
```
Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: Typo: missing a `y` in "Factory"; same for other annotations.

File: streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java [hunk 8]
```diff
@@ -1707,6 +1708,71 @@ public Integer call(String s) throws Exception {
...
+    // Function to create JavaStreamingContext without any output operations
+    // (used to detect the new context)
+    Function0<JavaStreamingContext> creatingFunc = new Function0<JavaStreamingContext>() {
+      public JavaStreamingContext call() {
+        return new JavaStreamingContext(conf, Seconds.apply(1));
+      }
+    };
+
+    ssc = JavaStreamingContext.getOrCreate(emptyDir.getAbsolutePath(), creatingFunc);
+    Assert.assertTrue("new context not created",
```
Comment (JoshRosen) [inline streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java]: Is your goal here to assert that `creatingFunc` was not called?  This seems like kind of a roundabout way of doing that.  I know that Mockito makes it pretty easy to write assertions that check the number of times that methods are invoked; alternatively, I guess you could just stick a "number of times called" counter in your function and check it in your assert.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala [hunk 9]
```diff
@@ -328,6 +330,138 @@ class StreamingContextSuite extends FunSuite with BeforeAndAfter with Timeouts w
     }
   }
 
+  test("getOrCreate") {
```
Comment (JoshRosen) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: This is a nice test; very comprehensive!

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala (hunk 5 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: Right. Damn, missed that.

File: streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala (hunk 9 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/StreamingContextSuite.scala]: Thank you!!

File: streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java (hunk 8 above)
Comment (tdas) [inline streaming/src/test/java/org/apache/spark/streaming/JavaAPISuite.java]: My knowledge of Mockito is limited, but it seemed to me that when you configure the mock with  `when...thenReturn` , you have to return a object, and you cannot on-demand call a function to create the object to return. That is necessary in this case, as you have to create a new StreamingContext when getOrCreate is called, and cannot configure with  `when...thenReturn` an already created context. 

That said, I think you are right about the current way be roundabout. I was trying to avoid having an extra flag var to signify called or not. But that would make it easier to understand. I will change it.

File: streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala (hunk 6 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StreamingContext.scala]: This is needed because there is already a version of getOrCreate with default arguments, and there can be only one overloaded version with default args.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala (hunk 7 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: Fixed. Good catch.

Comment (JoshRosen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: This still needs to be fixed on this line!

**Comment by JoshRosen** [comment]
LGTM pending Jenkins.

**Comment by tdas** [comment]
Merging this.

**Comment by zzcclp** [comment]
hi, @tdas , why this PR was be reverted?

**Comment by tdas** [comment]
This PR was reverted because I had used MutableBoolean which does not seem to work well with Hadoop 1.0.4. I reopened the PR in #5773.

**Comment by tdas** [comment]
I have revised the implementation of this PR in a followup PR #6096
